net-libs/nodejs: respect cpu_flags_x86_*

PF4Public opened the pull request.

PF4Public commented:
@williamh @thesamesam @mgorny

thesamesam commented:
@williamh could you take a look?

williamh commented:
I am concerned about this PR because it uses sed to patch the upstream source code conditionally. 
All upstream patches should be unconditional, so if you can make that happen, I'll take a look.
I recommend filing an issue [upstream](https://github.com/nodejs/node/issues) and discussing this with the nodejs authors directly.

PF4Public commented:
> it uses sed to patch the upstream source code conditionally

Why is it bad?

> All upstream patches should be unconditional

Why?

> if you can make that happen

How could one make several conflicting choices be unconditional?

thesamesam commented:
> > it uses sed to patch the upstream source code conditionally
> 
> Why is it bad?
> 

sed is indiscriminate.

> > All upstream patches should be unconditional
> 
> Why?
> 

Because you can't easily test it to see if it works with each combination and doesn't e.g. mangle the source code into something invalid.

See https://devmanual.gentoo.org/ebuild-writing/misc-files/patches/index.html#conditional-patching.

> > if you can make that happen
> 
> How could one make several conflicting choices be unconditional?

Ask them to make proper options available in the build system without sed mangling.

PF4Public commented:
> sed is indiscriminate.
> Because you can't easily test it to see if it works with each combination and doesn't e.g. mangle the source code into something invalid.

@thesamesam Both arguments are _generally_ correct, but in _this particular case_ I made sed very specific. First of all sed does not touch any of the sources, it targets `'sources'` section of a [gyp file](https://gyp.gsrc.io/docs/LanguageSpecification.md), which is a build configuration file, but if you consider it a source file too, so be it. The `'sources'` section _must_ exist. I then prepend this section with `'defines'` section followed by a comma, which is a valid syntax for this file. No failure possible. I can make sed `die` if it cannot find the needed section, but it would inevitably make [the expression less readable](https://github.com/PF4Public/gentoo-overlay/blob/fd7f5153b165819eb9eaa5f09feff9186b1c72b4/app-editors/vscode/vscode-9999.ebuild#L283-L284). Also maybe I shouldn't be **g**reedy with sed in this case. I can omit the **g**. 

I'd like to mention that both `if` and `else` cases have outcome, the result is absolutely deterministic, no mangling whatsoever! The result would look like this:
```sh
'defines': [ 'SIMDUTF_IMPLEMENTATION_ICELAKE=0', 'SIMDUTF_IMPLEMENTATION_HASWELL=1', 'SIMDUTF_IMPLEMENTATION_WESTMERE=1',], 'sources': [
```

So only four combinations: 1 1 1, 0 1 1, 0 0 1, 0 0 0. Which suggests it could count as non-conditional in the sense that `SIMDUTF_` defines are always present.

> See https://devmanual.gentoo.org/ebuild-writing/misc-files/patches/index.html#conditional-patching.

Quote from there: "_Patching is ideally only done to make the package in question build properly, and should not be done to modify the runtime behaviour of the package._" 
This patch does not alter runtime behaviour either. It merely excludes from compilation those functions, which wouldn't be used (on a particular CPU) anyway. This should also fix the issue where `-march=native` additionally sets `-mnoavx2` and similar. I remember there was such a WONTFIX issue in Bugzilla, but could not find it quickly to link here (probably it was filled for another package).

> Ask them to make proper options available in the build system without sed mangling.

While they do not pass defines to simdutf directly, passing defines via gyp seems to me kind of official or at least semi-official.

I hope I have addressed all your concerns, please let me know if there is anything else left.

PF4Public commented:
A month has passed. Anyone? @thesamesam @williamh

thesamesam commented:
The point I was making was more about why conditional patching, rather than unconditionally patching, is bad - but you've responded just about patching in general.

Anyway, it's not my decision. But if the code is already there and we're just mangling the 0 or 1 , it's fine with me and it sounds reasonable, especially given the issues we've had with it being clever in the past and failing to build.
